README
NAME
    Wight::Chart::Google - Save google charts to images using phantomjs

VERSION
    version 0.002

SYNOPSIS
      use Wight::Chart::Google;

      my $chart = Wight::Chart::Google->new(
        type => "area",
        width => 900,
        height => 500,
        options => {
          backgroundColor => 'transparent',
          hAxis => { gridlines => { color => "#fff" } },
          vAxis => { gridlines => { color => "#fff" } },
          legend => { position => 'none' },
        }
      );
      $chart->columns([
        { name => 'Day', type => 'string' },
        { name => 'Amount', type => 'number' },
      ]);
      $chart->rows([['1st',100], ['2nd',150], ['3rd',50], ['4th',70]]);
      $chart->render();

NAME
    Wight::Chart::Google - Generate static google charts

    This is pre-release software, everything could change and there are
    definitely bugs.

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/papercreatures/wight-chart-google/issues>. You will
    be notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/papercreatures/wight-chart-google>

      git clone git://github.com/papercreatures/wight-chart-google.git

AUTHOR
    Simon Elliott <simon@papercreatures.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Simon Elliott.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
0.002     2013-08-09 15:27:45 Europe/London

  decode UTF8, fix height option

0.001     2013-08-07 12:03:18 Europe/London

  Initial release


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;

use File::ShareDir::Install;
install_share dist => "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Save google charts to images using phantomjs",
  "AUTHOR" => "Simon Elliott <simon\@papercreatures.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "File::ShareDir::Install" => "0.03"
  },
  "DISTNAME" => "Wight-Chart-Google",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Wight::Chart::Google",
  "PREREQ_PM" => {
    "Cwd" => 0,
    "Encode" => 0,
    "File::Share" => 0,
    "JSON::XS" => 0,
    "Moo" => 0,
    "Wight" => 0,
    "strictures" => 1
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "List::Util" => 0,
    "Test::More" => "0.88",
    "blib" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}




lib/Wight/Chart/Google.pm
package Wight::Chart::Google;

#ABSTRACT: Save google charts to images using phantomjs

our $VERSION = '0.002'; # VERSION

use strictures 1;
use Moo;
use Wight;
use Encode;
use File::Share qw/dist_file/;
use Cwd;
use JSON::XS;

#TODO: import roles for each type for more options
my $types = {
  line => "LineChart",
  area => "AreaChart",
  bar => "BarChart",
  spark => 'ImageSparkLine',
};

has 'type' => ( is => 'rw', required => 1, isa => sub { $types->{$_[0]} } );
has 'output' => ( is => 'rw', default => sub { 'chart.png'} );
has 'rows' => ( is => 'rw' );
has 'options' => ( is => 'rw', default => sub { {} } );
has 'columns' => ( is => 'rw' );
has 'wight' => ( is => 'lazy' );
has 'width' => ( is => 'rw', default => 900 );
has 'height' => ( is => 'rw', default => 500 );

sub _build_wight {
  my $self = shift;
  my $wight = Wight->new();
  #really?
  my $file = -e 'share/chart.html' ?
    getcwd . '/share/chart.html'
    :
    dist_file('Wight-Chart-Google', 'chart.html');


  $wight->resize($self->width + 20, $self->height + 20);
  $wight->visit("file:///$file");
  $wight;
}

sub render {
  my ($self) = @_;
  my $w = $self->wight;
  my $options = {
    chartArea => {
      width => $self->width - 100,
      height => $self->height - 100,
    },
    %{$self->options}
  };

  my $args = decode_utf8(encode_json({
    options => $options,
    type => $types->{$self->type},
    rows => $self->rows,
    columns => $self->columns,
  }));
  #warn $args;
  $w->evaluate("drawChart($args)");
  $w->render($self->output);
}


1;

__END__

=pod

=encoding utf-8

=head1 NAME

Wight::Chart::Google - Save google charts to images using phantomjs

=head1 VERSION

version 0.002

=head1 SYNOPSIS

  use Wight::Chart::Google;

  my $chart = Wight::Chart::Google->new(
    type => "area",
    width => 900,
    height => 500,
    options => {
      backgroundColor => 'transparent',
      hAxis => { gridlines => { color => "#fff" } },
      vAxis => { gridlines => { color => "#fff" } },
      legend => { position => 'none' },
    }
  );
  $chart->columns([
    { name => 'Day', type => 'string' },
    { name => 'Amount', type => 'number' },
  ]);
  $chart->rows([['1st',100], ['2nd',150], ['3rd',50], ['4th',70]]);
  $chart->render();

=head1 NAME

Wight::Chart::Google - Generate static google charts

This is pre-release software, everything could change and there are definitely bugs.

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/papercreatures/wight-chart-google/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/papercreatures/wight-chart-google>

  git clone git://github.com/papercreatures/wight-chart-google.git

=head1 AUTHOR

Simon Elliott <simon@papercreatures.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Simon Elliott.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
